### PR TITLE
fix: race condition in keploy record

### PR DESCRIPTION
**Pull request by dxtym**
## Related Issue
  - Race condition with `testCount` while exiting keploy record 

Closes: #1851

#### Describe the changes you've made
Put `sync.Mutex` locks over the sharing value so that concurrent activities don't cause race condition. Add `sync.waitGroup` to control the flow of actions between the main and sub gorountines.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added

No additional tests. Only the build procedure at #1851 

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
NA

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

|        Original         |          Updated           |
|:-----------------------:|:--------------------------:|
| image | image |

### CODE REVIEW

**Comment by dxtym** [comment]
Singed-off-by: Dilmurod Abdusamadov dilmurod.abdusamadov2004@gmail.com

**Comment by dxtym** [comment]
I have read the CLA Document and I hereby sign the CLA

File: pkg/service/record/record.go [hunk 1]
```diff
@@ -66,13 +67,18 @@ func (r *Recorder) Start(ctx context.Context) error {
...
+	var waitGroup = &sync.WaitGroup{}
 	var mockCountMap = make(map[string]int)
 
 	// defering the stop function to stop keploy in case of any error in record or in case of context cancellation
 	defer func() {
 		select {
 		case <-ctx.Done():
+			waitGroup.Wait()
+			mutex.Lock()
 			r.telemetry.RecordedTestSuite(newTestSetID, testCount, mockCountMap)
```
Comment (slayerjain) [inline pkg/service/record/record.go:80]: I might be missing something but not really sure if this defer could be called concurrently 🤔

Comment (slayerjain) [reply pkg/service/record/record.go:80]: aah, so its the testCount probably..

Comment (slayerjain) [reply pkg/service/record/record.go:80]: I think in the defer function if we wait for the other goroutines to complete/cancel - this race condition can be better resolved. Maybe we can move the `runAppErrGrp.Wait()` in the defer to the top. 

@charankamarapu @gouravkrosx what do you think?

**Review by slayerjain — CHANGES_REQUESTED** [review]
Hi @dxtym , Thanks for the PR! I have left some comments and I think @gouravkrosx / @charankamarapu should give more context for a better solution.

File: pkg/service/record/record.go (hunk 1 above)
Comment (gouravkrosx) [reply pkg/service/record/record.go:80]: @dxtym , we can move the `r.telemetry.RecordedTestSuite(newTestSetID, testCount, mockCountMap)` to the bottom of the defer, this way we are waiting for all the go routines to complete their execution.

Comment (dxtym) [reply pkg/service/record/record.go:80]: Thank you for the comments @slayerjain @gouravkrosx! I moved the `r.telemetry.RecordedTestSuite(newTestSetID, testCount, mockCountMap)` to the bottom and removed mutex and waitGroups. Did I get your feedback right?

Comment (slayerjain) [reply pkg/service/record/record.go]: @dxtym yes, I think you've made the changes according to the feedback. 

It seems like `ctx` cancellation is not working correctly based on the this pipeline - https://github.com/keploy/keploy/actions/runs/9034964803/job/24829105617?pr=1867


It seems to be stuck on the last line: 

Screenshot 2024-05-10 at 9 55 22 PM

Comment (dxtym) [reply pkg/service/record/record.go]: image
I checked one of the previous PRs that passed all the tests. It had the same line. Could it be something else?

Comment (dxtym) [reply pkg/service/record/record.go]: What if I undo the changes to `ctx` cancelling part by not moving down? I did the previous solution with `mutex` while removing  the `waitGroup`, and no race condition occurred.  Perhaps, we could go with that )

Comment (dxtym) [reply pkg/service/record/record.go]: I tried removing `waitGroup` from my previous solution, and the binary worked the same way. Perhaps we could try that )

**Comment by slayerjain** [comment]
@gouravkrosx please help close this

**Comment by dxtym** [comment]
> @gouravkrosx please help close this

I would greatly appreciate support 🙏

**Comment by gouravkrosx** [comment]
@dxtym, Sorry for the delay, will try to review this PR before EOD.

**Comment by gouravkrosx** [comment]
@dxtym , In this [commit](https://github.com/keploy/keploy/pull/1867/commits/708485c049e675231d9eaa21dc01f79e26a2ad06), you not only moved the `r.telemetry.RecordedTestSuite(newTestSetID, testCount, mockCountMap)` but also moved the **_select_** statement which was listening to context done. That made this [pipeline](https://github.com/keploy/keploy/actions/runs/9034964803/job/24829105617?pr=1867) stuck.
You just need to move this `r.telemetry.RecordedTestSuite(newTestSetID, testCount, mockCountMap)` at the bottom outside the **_select_** scope but inside the **_defer_** function scope. And that should work.

**Comment by dxtym** [comment]
@gouravkrosx Thank you for your guidance. Made some changes according to the feedback.

File: pkg/service/record/record.go [hunk 2]
```diff
@@ -72,13 +71,14 @@ func (r *Recorder) Start(ctx context.Context) error {
 	defer func() {
 		select {
 		case <-ctx.Done():
-			r.telemetry.RecordedTestSuite(newTestSetID, testCount, mockCountMap)
+			break
```
Comment (gouravkrosx) [inline pkg/service/record/record.go:74]: This break is not required as the select statement is not in a for loop.

**Comment by gouravkrosx** [comment]
@dxtym Please resolve the addressed comments and please fix the lint pipeline.

File: pkg/service/record/record.go (hunk 2 above)
Comment (dxtym) [reply pkg/service/record/record.go:74]: Can I open another PR for this issue and close this one? It looks like my local copy became a bit of a mess. I really apologize for that.

**Comment by gouravkrosx** [comment]
Closed as contributor made new PR: https://github.com/keploy/keploy/pull/1914

**Pull request closed without merging**